FEA add PredictionErrorDisplay

glemaitre opened the pull request:
Add a `PredictionErrorDisplay` which is useful when dealing with regression.

Edit: Partially addresses #16608.

glemaitre committed:
FEA add plot_prediction_error

glemaitre committed:
iter

glemaitre committed:
iter

glemaitre committed:
Merge remote-tracking branch 'origin/master' into predicted_actual_regression_plot

glemaitre committed:
iter

glemaitre committed:
check with an example

glemaitre committed:
reset previous style

glemaitre committed:
iter

glemaitre committed:
change another example

glemaitre committed:
DOC update another example

glemaitre committed:
PEP8

NicolasHug commented:
i'm getting baited into reviewing this but it doesn't seem finished yet... mark as WIP @glemaitre ? ;)

glemaitre commented:
Ups sorry. This is WIP :)

glemaitre commented:
You can have a look at the API thought. I change examples and I am writing proper test now that I did my example-based development :)

glemaitre committed:
add some tests

glemaitre committed:
TST add test and improve documentation

glemaitre commented:
@NicolasHug @thomasjpfan I think the PR is ready to get some attention.
Open to any suggestion regarding the AP

glemaitre committed:
compatibility matplotlib 2.2

NicolasHug commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,356 @@
+
+        random_state = check_random_state(self.random_state)
+
+        subsample = self.subsample
```
shouldn't the Display exclusively focus on plotting logic? I feel like anything compute-related should be done outside of the display

glemaitre replied:
True

NicolasHug commented on examples/compose/plot_transformed_target.py:
```diff
@@ -16,28 +16,13 @@
 # %%
+print(__doc__)
+
+# %% [markdown]
```
I'm curious as to what the `[markdown]` tags are used for?

glemaitre replied:
It is a jupytext marker to show the paragraph when executing the script in some IDE (like VS code). True that this might not be necessary.

NicolasHug reviewed (commented):
Thanks @glemaitre,

I only took a very brief look but my impression is that this might be a bit too high-level for scikit-learn.

My understanding of our goal with the plotting utilities is that we allow users to i) compute complicated stuff like PDPs and ii) update the plot styles without having to re-compute the results every time.

It seems  that error plots are really just a matter of calling `predict` and `scatter` so it's already quite simple.

Just my personal thoughts, no strong opinion anyway

glemaitre commented:
> It seems that error plots are really just a matter of calling predict and scatter so it's already quite simple.

I agree with these statements. Basically, I wanted my arguments were the following:

* it is a useful plot that you want to use often in teaching (we need to it during developing our MOOC) or even in practice.
* the predict and scatter are actually the easy part. What I most dislike with some plots that I saw around: (i) the wrong diagonal is shown and (ii) the aspect ratio of the plot is not squared.

So I agree that this is more of a convenience function.

glemaitre committed:
avoid computation in display class

glemaitre committed:
EXA remove subsample from example

jnothman commented:
And I like its convenience! I am not sure we need to limit our plotting utilities to things that are difficult to achieve without.

NicolasHug commented on examples/ensemble/plot_stack_predictors.py:
```diff
@@ -188,62 +188,40 @@ def load_ames_housing():
+        for key, value in scorers.items()
+    }
+
+    display = PredictionErrorDisplay(y_true=y, y_pred=y_pred, scores=scores)
```
do we really want to directly use `PredictionErrorDisplay` in this example (which is not about advanced plotting stuff)? It seems contradictory to our recommendation to rather rely on `plot_prediction_error`

glemaitre replied:
Here I am doing this because we are computed the scores via cross-validation while the the `plot_*` function would compute for the set passed as argument. Another way would be to not call `plot_*` and then affect the legend to the axis manually.  I will make a try to see if this is readable

NicolasHug commented on examples/inspection/plot_linear_model_coefficient_interpretation.py:
```diff
@@ -158,39 +168,38 @@
+    "MAE on testing set": f"{mae_test:.2f} $/hour"
+}
+
+display = PredictionErrorDisplay(y_true=y_test, y_pred=y_pred, scores=scores)
```
same comment here: shouldn't we be using `plot_prediction_error`?

NicolasHug commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,336 @@
+        Fitted regressor or a fitted :class:`~sklearn.pipeline.Pipeline`
+        in which the last estimator is a regressor.
+
+    X : {array-like, sparse matrix} of shape (n_samples, n_features)
+        Input values.
+
+    y : array-like of shape (n_samples,)
+        Target values.
```
in relation to my previous comments, I'm wondering whether it really makes sense to pass X and y here. Maybe it should be `y` and `y_pred` instead: at least according to our examples, having `y_pred` already computed seems to be the most common use-case.

NicolasHug replied:
Or maybe we should accept both strategies...

Some related discussion in https://github.com/scikit-learn/scikit-learn/issues/15880

glemaitre replied:
Yes, we encounter this issue where we would like to use precomputed prediction and score.
We actually seem to not have to figure out the API that we want to go with. I don't know if would be safer to restrict the current signature and make a move afterwards.

NicolasHug replied:
CC @thomasjpfan @amueller 

One difference here vs #15880 is that we don't have to worry about backward compat, and we can come up with whatever API we want/need.


It seems to me that allowing both `estimator, X` and `y_pred` would be the most natural solution here (while keeping them mutually exclusive).

glemaitre replied:
Having a single first tuple argument with `(estimator, X, y)` or `(y_true, y_pred)` would be nice in this case.
You can switch easily by detecting in which case you are and might not be that user-unfriendly.

glemaitre committed:
Merge remote-tracking branch 'origin/main' into predicted_actual_regression_plot

glemaitre committed:
iter

glemaitre committed:
iter

glemaitre committed:
iter

glemaitre committed:
iter

glemaitre committed:
iter

glemaitre committed:
iter

glemaitre committed:
iter

glemaitre committed:
docstring

glemaitre committed:
docstring

lorentzenchr commented:
@glemaitre Could you merge main so that the doc is rendered again?

glemaitre committed:
Merge remote-tracking branch 'origin/main' into predicted_actual_regression_plot

glemaitre committed:
Merge remote-tracking branch 'origin/main' into predicted_actual_regression_plot

glemaitre committed:
revert pre-commit change

glemaitre committed:
black

glemaitre committed:
iter

lorentzenchr commented:
I like such a plotting functionality very much. If we add - in a later PR - a binning strategy, this could become the equivalent of the reliability diagram for regression.

glemaitre commented:
> I like such a plotting functionality very much. If we add - in a later PR - a binning strategy, this could become the equivalent of the reliability diagram for regression.

Nice. IRL with @ogrisel and @jeremiedbb we were wondering if there are other types of plots that are useful in regression settings with Poisson or Gamma distribution. 
@lorentzenchr did you encounter any other types of plots for those specific use cases.

It would help at knowing if the current API is right.

glemaitre committed:
improve example and change API

glemaitre committed:
doc fix

glemaitre committed:
fix tests

glemaitre committed:
TST add residuals test

glemaitre committed:
fixes

lorentzenchr commented:
> did you encounter any other types of plots for those specific use cases?

`PredictionErrorDisplay` offeres several plots and those go under several names, e.g.
-  actual vs expected
- residual plot

One point is, that the actual implementation is focused on modelling the conditional expectation, E[Y|X], only. It is not suited as diagnostics for models predicting a quantile. This is, maybe, the most critical API consideration to be done.

I know of https://arxiv.org/abs/2108.03210 that generalizes reliability diagrams to "T-reliability diagrams" for any identifiable funtional T (like the expectation and quantiles).

glemaitre committed:
Merge remote-tracking branch 'origin/main' into predicted_actual_regression_plot

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/glemaitre/18020

jeremiedbb committed:
target 1.2

lorentzenchr commented:
Could we flit x- and y-axis such that predicted values are on the x-axis and observed values on the y-axis? This way, the x-axis would be the same as for the residuals.

jeremiedbb committed:
Merge branch 'main' into predicted_actual_regression_plot

jeremiedbb commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,454 @@
+        if errors_kwargs is None:
+            errors_kwargs = {}
+
+        default_scatter_kwargs = {"color": "black", "alpha": 0.8}
+        default_line_kwargs = {
+            "color": "tab:blue",
```
I find that having the line in blue and the dots in black make the plot a bit sad. What do you think about switching and plot the dots in blue and the line in black (or red) ?

jeremiedbb commented on examples/compose/plot_transformed_target.py:
```diff
@@ -73,91 +69,93 @@
 # prediction. Subsequently, a logarithmic function is used to linearize the
 # targets, allowing better prediction even with a similar linear model as
 # reported by the median absolute error (MAE).
+from sklearn.metrics import mean_absolute_error, r2_score
+
+
+def compute_score(y_true, y_pred):
+    return {
+        "R2 score": f"{r2_score(y_true, y_pred):.3f}",
+        "MAE": f"{mean_absolute_error(y_true, y_pred):.3f}",
+    }
```
The example was using the median_absolute_error. Is it intended to use the mean_absolute_error ? if so, references to median_absolute_error should be replaced. On a side note, it looks like MAE usually means mean_abs_error while median_abs_error is usually abreviated MedAE.

jeremiedbb reviewed (commented):
Even if it's for a simple plot, I like its convenience as well. And it could be extended in the future to plot a little bit more advanced stuff, like an histogram of the residuals on the side for instance.

Here are some comments

jeremiedbb commented:
> Could we flit x- and y-axis such that predicted values are on the x-axis and observed values on the y-axis? This way, the x-axis would be the same as for the residuals.

I took a look at what's usually done and it's almost always observed vs actual. In particular, it's how we used to do in our examples. So I think it's better from a user point of view to keep the existing / expected behavior to avoid surprises.

lorentzenchr commented:
I find it bad UX to have the same function put the same thing (predicted values) on 2 different axis depending on the `kind` parameter. This is particular so with the two plots next to each other.

jeremiedbb commented:
@glemaitre this is the plot that we speaked about irl, with a histogram distribution of the residuals on the side that I found interesting
![34059Residual_lasso](https://user-images.githubusercontent.com/34657725/192746794-81ef9c78-ee72-4801-9956-1971e1a58336.jpg)
I don't think it needs to be done as part of this PR but I think it's worth keeping in mind for a later PR once this one is merged

glemaitre commented:
> I find it bad UX to have the same function put the same thing (predicted values) on 2 different axis depending on the kind parameter. This is particular so with the two plots next to each other.

I will go back to this PR. I am wondering if we could have a keyword for this option. A vast majority of plots are using the representation where the x-axis is the true value. We can keep this behaviour as default but offer the option for the case presented by @lorentzenchr

glemaitre committed:
Merge remote-tracking branch 'origin/main' into predicted_actual_regression_plot

glemaitre committed:
iter

glemaitre commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,454 @@
+    y_pred : ndarray of shape (n_samples,)
+        Prediction values.
+
+    scores : dict, default=None
+        Dictionary where the key is the name of the metric displayed and the
+        value is the metric value.
```
I would argue that this is not that easy to do:

```python
            extra = plt.Rectangle(
                (0, 0), 0, 0, fc="w", fill=False, edgecolor="none", linewidth=0
            )
            scoring_legend = "\n".join(
                [f"{name} = {value}" for name, value in self.scores.items()],
            )
            ax.legend([extra], [scoring_legend])
```

glemaitre replied:
Having a second thought about this one. We can remove it and have in the example how to plot it. I can imagine that some people want to have the info in the title for instance.

glemaitre committed:
switch color

glemaitre committed:
add horizontal line for residuals plot

glemaitre committed:
add horizontal line for residuals plot

glemaitre commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,454 @@
+            Dictionary with keyword passed to the `matplotlib.pyplot.plot`
+            call to draw the diagonal line.
+
+        errors_kwargs : dict, default=None
+            Dictionary with keyword passed to the `matplotlib.pyplot.plot`
+            call to draw the residual lines. Only taken into account when
+            `with_errors=True`.
```
Actually, we are not using this feature in scikit-learn. It should be something that I used in the MOOC of scikit-learn but for pedagogical purposes. I will remove this item for the moment.

glemaitre committed:
update style regarding scores

glemaitre committed:
Add the option to swap x- y-axsi

glemaitre commented:
I added a keyword `x_axis` to swap the axis. We use it in the example when plotting the residuals and predictions on the same plot. Indeed, it was already the case on the current example.

lorentzenchr commented:
It's not all that simple:
- Residual (y) vs predicted (x) seems common ground. But note that it only makes sense if predicting the mean E(Y|X), it is nonsense (in this form, observed - predicted) for quantile regression.
- Actual/Observed (y) vs predicted (x): (I know it in this form and could not quickly find a book that does it otherwise.) If you aggregate the actual/observed in a (statistically sound) way per predicted value, you get a reliability diagram / calibration plot.

glemaitre committed:
DOC add some more documentation

glemaitre committed:
Merge remote-tracking branch 'origin/main' into predicted_actual_regression_plot

glemaitre commented:
OK. I am pretty happy with the current stage. I could also improve the documentation and remove useless keywords. I think that I addressed the concern of both @jeremiedbb and @lorentzenchr

Do you want to make an additional review in hopes it goes in 1.2 ;)

glemaitre committed:
DOC outdated versionadded

anntzer committed:
DOC Clarify sorting order of PCA components (#24531)

Dev-Khant committed:
DOC Mention pandas dataframe support in ColumnTransformer in FAQ (#24532)

glemaitre committed:
MAINT modify import raccoon face for SciPy >= 1.10 (#24521)

thomasjpfan committed:
CI Remove unneeded setup-python (#24544)

thomasjpfan committed:
FIX Treat gradient boosting categoricals outside the bounds as unknown during predict (#24283)

adrinjalali committed:
MNT update security.md (#24551)

thomasjpfan committed:
BLD Import from numpy/arrayobject.h directly (#24547)

thomasjpfan committed:
FIX Fixes HTML repr when get_params contains classes (#24512)



Co-authored-by: Adrin Jalali <adrin.jalali@gmail.com>

rhelmeczi committed:
MAINT remove redundant items from iterable (#24554)

jeremiedbb committed:
MAINT Clean deprecation for 1.2: get_feature_names (#24395)

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

rjuer committed:
DOC Ensure that sklearn.utils.axis0_safe_slice passes numpydoc (#24561)

mauroantonioserrano committed:
DOC Ensures that sklearn.tree._export.plot_tree passes numpydoc validation (#24559)

Co-authored-by: Juan Martin Loyola <jmloyola@outlook.com>
Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rjuer committed:
DOC Ensure that sklearn.utils.is_scalar_nan passes numpydoc validation (#24562)

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rprkh committed:
DOC make `plot_mean_shift.py` more colourblind friendly (#24553)

lorentzenchr committed:
ENH change Ridge tol to 1e-4 (#24465)

* ENH set defaul tol of Ridge to 1e-4 instead of 1e-3

* MNT rename internal least square method in discrimant_analysis

* DOC add versionchanged

* DOC add whats_new entry

* DOC add pr to whatsnew

* DOC better versionchanged message

thomasjpfan committed:
MNT Optimize safe_indexing for slices (#24587)

adrinjalali committed:
CI remove LGTM (#24592)

thomasjpfan committed:
CLN Remove unneeded variable definition in DictVectorizer.fit (#24590)

irene000 committed:
DOC Ensures that available_if passes numpydoc validation (#24586)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

thomasjpfan committed:
CLN Migrate avaliable_if to it's own file (#24594)

jeremiedbb commented on doc/modules/model_evaluation.rst:
```diff
@@ -2694,6 +2694,38 @@ Here are some usage examples of the :func:`d2_absolute_error_score` function::
+there is a relationship between the value of the target and the residuals, e.g.
+the error increases when the true target increases.
+
+Refer to the example below to see an model evaluation that make use of this
```
```suggestion
Refer to the example below to see a model evaluation that make use of this
```

jeremiedbb commented on doc/modules/model_evaluation.rst:
```diff
@@ -2694,6 +2694,38 @@ Here are some usage examples of the :func:`d2_absolute_error_score` function::
+Visual evaluation of regression models
+--------------------------------------
+
+The :class:`~sklearn.metrics.PredictionErrorDisplay` displays allows to visually
+inspect the quality of regression models. The quality of a regression model can
```
Maybe mention that it's not applicable for quantile regression as Christian pointed out

jeremiedbb reviewed (approved):
LGTM !

clytaemnestra committed:
API Deprecate the extra keyword arguments of utils.extmath.density (#24523)

Co-authored-by: Meekail Zain <34613774+Micky774@users.noreply.github.com>
Co-authored-by: Jérémie du Boisberranger <34657725+jeremiedbb@users.noreply.github.com>

thomasjpfan committed:
CLN Do not override signature for _visual_block_ (#24588)

mansi1597 committed:
TST Removed sklearn.util.fixes.linspace from numpydoc ignore list (#24582)

awinml committed:
DOC Ensures that sklearn.utils.extmath.safe_sparse_dot passes numpydoc validation (#24567)

RowanMankoo committed:
CLN Remove unnecessary operation in mutual_info (#24569)

jjerphan committed:
TST Relax `test_gradient_boosting_early_stopping` (#24541)

awinml committed:
DOC Ensures that sklearn.utils.extmath.weighted_mode passes numpydoc validation (#24571)

Co-authored-by: jeremie du boisberranger <jeremiedbb@yahoo.fr>

lorentzenchr commented:
I really do not intend to block this PR, but I'm not at ease in it's current form. Could we please start minimal with only the plot of the residuals?

jeremiedbb committed:
TST Make test_function_docstrings ignore functions from utils.fixes (#24602)

jeremiedbb committed:
MAINT Clean deprecation for 1.2: cv_results_ keys (#24599)

jeremiedbb committed:
MAINT Clean deprecation for 1.2: load_boston (#24603)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

thatgeeman committed:
DOC Ensure that gen_even_slices passes numpydoc validation (#24608)

ogrisel committed:
DOC Check sha256 digests of tarballs in tutorial and examples before extraction (#24617)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

ogrisel committed:
MAINT set joblib min to 1.1.1 (#24621)

Micky774 committed:
API Add `"auto"` value and deprecate default value for `normalized_stress` (#23834)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

mansi1597 committed:
DOC Ensures that svd_flip passes numpydoc validation (#24581)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

lorentzenchr committed:
ENH FEA add interaction constraints to HGBT (#21020)

Co-authored-by: Loïc Estève <loic.esteve@ymail.com>

glemaitre committed:
DOC add a docstring example for the learning_curve function (#24546)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jeremiedbb committed:
MAINT Clean deprecation for 1.2: normalize in linear models (#24391)

glemaitre commented:
> I really do not intend to block this PR, but I'm not at ease in it's current form.

Since you are the only second reviewer, it is equivalent to a stalled PR :)

> Could we please start minimal with only the plot of the residuals?

Let's make the change then.

glemaitre commented:
> Could we please start minimal with only the plot of the residuals?

Actually, I am confused. I would have started for the observed vs. actual thinking that there is no real controversy there, isn't it?

jjerphan committed:
Simplify Tempita preprocessing (#24624)

jjerphan committed:
GEMMTermComputer.{_compute_distances_on_chunks→_compute_dist_middle_terms} (#24626)

lorentzenchr commented:
Consider the following plots:

#### 1. residuals
![image](https://user-images.githubusercontent.com/15324633/195443783-a0625263-e800-4356-bc2c-e30b30eafa08.png)
You don't see much, a heatmap might be better, or boxplots with binning. It could be, that the mean residual is approximately zero for every predicted value.

#### 2. predicted vs actual version
![image](https://user-images.githubusercontent.com/15324633/195441850-e9d361b0-09af-493e-bf9a-99797f81298a.png)
This does not help in any way.

#### 3. predicted vs true conditional expectation E[Y|X]
What would help, but is usually unavailable/unfeasible is predicted vs true expected value
![image](https://user-images.githubusercontent.com/15324633/195442044-a8afc685-f6fa-4aeb-9e94-80ee2d92bfc3.png)

#### 4. reliability diagram for regression
What actually helps and is available is a reliability diagram for regression (target=expectation):
![image](https://user-images.githubusercontent.com/15324633/195443286-55037a20-f3d7-4b07-ba0b-5d6a3bf0b39d.png)

<details>

```python
import matplotlib.pyplot as plt
import numpy as np
from sklearn.isotonic import IsotonicRegression
from sklearn.linear_model import GammaRegressor


n = 40_000
rng = np.random.default_rng(973)
X = rng.uniform(-1, 1, size=(n, 1))
y = rng.lognormal(mean=X.squeeze())  # E[Y|X] = exp(X + 1/2)
def mean_of_y(X):
    return np.exp(X + 1/2)
X_train, X_test = X[:n//2], X[n//2:,:]
y_train, y_test = y[:n//2], y[n//2:]
model = GammaRegressor().fit(X_train, y_train)
# %%
plt.scatter(model.predict(X_test), y_test - model.predict(X_test))
plt.plot(model.predict(X_test), np.zeros_like(y_test), 'k--')
plt.xlabel("predicted values")
plt.ylabel("residuals")
plt.title("Residual plot")
# %%
plt.scatter(y_test, model.predict(X_test))
plt.plot(y_test, y_test, 'k--')
plt.xlabel("actual values")
plt.ylabel("predicted values")
plt.title("Predicted vs Observed")
# %%
plt.scatter(mean_of_y(X_test), model.predict(X_test))
plt.plot(mean_of_y(X_test), mean_of_y(X_test), 'k--')
plt.xlabel("true mean E[Y|X]")
plt.ylabel("predicted values")
plt.title("Predicted vs True Expectation")
# %%
idx = np.argsort(model.predict(X_test))
predicted = model.predict(X_test)[idx]
observed = y_test[idx]
plt.plot(predicted, IsotonicRegression().fit(predicted, observed).predict(predicted))
plt.plot(predicted, predicted, 'k--')
plt.xlabel("prediction")
plt.ylabel("estimate of E[Y|prediction]")
plt.title("Reliability diagram for regression")
```

<(details>

thomasjpfan committed:
ENH Introduces set_output API for pandas output (#23734)

* Introduces set_output API for all transformers

* TransformerMixin inherits from _SetOutputMixin

* Adds tests

* Adds whatsnew

* Adds example on using set_output API

* Adds developer docs for set_output

thomasjpfan committed:
CI Use GITHUB_OUTPUT instead of deprecated set-output (#24644)

* CI Use GITHUB_ENV instead of deprecated set-output

* CI [cd build gh]

* FIX Echo into github environment [cd build gh]

* FIX Use GITHUB_OUTPUT [cd build gh]

jjerphan committed:
DOC Promote Meekail Zain to the Core Contributor Team (#24649)

thomasjpfan committed:
CI Remove Windows 32 bit support (#24627)

ArturoAmorQ committed:
DOC Use :doi: directive for KMeans (#24641)

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

ArturoAmorQ committed:
DOC Rework plot_roc.py example (#24200)

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>
Co-authored-by: Chiara Marmo <cmarmo@users.noreply.github.com>
Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jeremiedbb committed:
MAINT clean deprecation for 1.2: leftovers (#24647)

thatgeeman committed:
DOC Ensure that gen_batches passes numpydoc validation (#24609)

jjerphan committed:
MAINT `PairwiseDistancesReduction`: Update comments and remove unused symbols (#24625)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

lesteve committed:
MAINT Fix full doc build by avoiding plot_set_output.py side-effect (#24654)

jeremiedbb committed:
MAINT Clean deprecation for 1.2: load_boston follow-up (#24653)

* remove load_boston references from the docs

* iter

* update poor_score

ArturoAmorQ committed:
MAINT set plotly min to 5.10 (#24629)

ogrisel committed:
OPTIM use pairwise_distances_argmin in NearestCentroid.predict (#24645)



Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

glemaitre committed:
MAINT use nanmin to replace nan by finite values in ranking of SearchCV (#24543)


Co-authored-by: Tim Head <betatim@gmail.com>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>
Co-authored-by: Loïc Estève <loic.esteve@ymail.com>

jjerphan committed:
MAINT `PairwiseDistancesReduction`: Rename some symbols and files (#24623)

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>
Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

awinml committed:
DOC Ensures that `sklearn.utils.extmath.randomized_svd` passes numpydoc validation (#24607)

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

michpara committed:
DOC Ensures that if_delegate_has_method passes numpydoc validation (#24633)

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>
Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

cmarmo committed:
Add msvcp140.dll to Windows 64 bit wheels (#24631)

awinml committed:
DOC Ensures that sklearn.utils.extmath.fast_logdet passes numpydoc validation (#24605)

Co-authored-by: jeremie du boisberranger <jeremiedbb@yahoo.fr>

thomasjpfan committed:
MNT Do not update docs with deprecated decorator (#24410)


Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jeremiedbb committed:
MAINT Clean deprecation for 1.2: default random_state in randomized_svd (#24661)

jeremiedbb committed:
MAINT Clean deprecation for 1.2: feature names exact match (#24660)

jeremiedbb committed:
MAINT Bump min dependencies for 1.2 (#24650)


Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

rusdes committed:
API Remove `sklearn.metrics.manhattan_distances` option `sum_over_features` (#24630)

jeremiedbb committed:
EFF avoid computing inertia in KMeans' predict (#24666)

rprkh committed:
DOC make `plot_agglomerative_clustering_metrics.py` colorblind friendly (#24655)

x110 committed:
DOC use KBinsDiscretizer in lieu of KMeans in vector quantization example (#24374)

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>
Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

OmarManzoor committed:
TST use global_random_seed in sklearn/cluster/tests/test_dbscan.py (#24598)

lesteve committed:
HOTFIX Temporarily disable py38_conda_defaults_openblas build (#24693)

betatim committed:
DOC Fix typo and adjust wording in `set_output` example (#24689)

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

thomasjpfan committed:
DOC Improve docstring around set_output (#24672)

* DOC Improve docstring around set_output

* DOC Improve docs around set_output

* DOC Address comments

* DOC Better grammar

* DOC Improve wording

* DOC Improves docstring in set_config

thomasjpfan committed:
ENH Makes OneToOneFeatureMixin and ClassNamePrefixFeaturesOutMixin public (#24688)

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

AlessandroMiola committed:
DOC Fix typo in plot_set_output.py example (#24704)

lesteve committed:
MAINT Fix build when SKLEARN_OPENMP_PARALLELISM_ENABLED=False (#24682)

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

glemaitre committed:
MAINT renable Linux + Python 3.8 build with OpenBLAS (#24705)

cmarmo committed:
CI Add wheel builds for Python 3.11 (#24446)

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>
Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

thomasjpfan committed:
CI Remove remaining windows 32 references (#24657)

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

zaznaczony committed:
DOC fix typo inside Pipeline docstring (#24730)

cmarmo committed:
DOC fix title underline too short in Gaussian Process kernel (#24726)

ftorres16 committed:
DOC correct bound of sum LinearSVR in formula in svm.rst (#24722)

glemaitre committed:
DOC fix sphinx directive in  function (#24733)

glemaitre committed:
DOC fix deprecation warning raised by KMeans and Matplotlib (#24692)

Co-authored-by: Tim Head <betatim@gmail.com>

cmarmo committed:
Add sphinx_highlight.js to the search page (needed since sphinx 5.2.0). (#24727)

glemaitre committed:
DOC fix a missing final fullstop in docstring (#24739)

ArturoAmorQ committed:
DOC Improve narrative of plot_roc_crossval example (#24710)

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

lorentzenchr committed:
FEA add (single) Cholesky Newton solver to GLMs (#24637)

* FEA add NewtonSolver, CholeskyNewtonSolver and QRCholeskyNewtonSolver

* ENH better singular hessian special solve

* CLN fix some typos found by reviewer

* TST assert ConvergenceWarning is raised

* MNT add BaseCholeskyNewtonSolver

* WIP colinear design in GLMs

* FIX _solve_singular

* FIX false unpacking in

* TST add tests for unpenalized GLMs

* TST fix solutions of glm_dataset

* ENH add SVDFallbackSolver

* CLN remove SVDFallbackSolver

* ENH use gradient step for singular hessians

* ENH print iteration number in warnings

* TST improve test_linalg_warning_with_newton_solver

* CLN LinAlgWarning fron scipy.linalg

* ENH more robust hessian

* ENH increase maxls for lbfgs to make it more robust

* ENH add hessian_warning for too many negative hessian values

* CLN some warning messages

* ENH add lbfgs_step

* ENH use lbfgs_step for hessian_warning

* TST make them pass

* TST tweek rtol for lbfgs

* TST add rigoros test for GLMs

* TST improve test_warm_start

* ENH improve lbfgs options for better convergence

* CLN fix test_warm_start

* TST fix assert singular values in datasets

* CLN address most review comments

* ENH enable more vebosity levels for lbfgs

* DOC add whatsnew

* CLN remove xfail and clean a bit

* CLN docstring about minimum norm

* More informative repr for the glm_dataset fixture cases

* Forgot to run black

* CLN remove unnecessary filterwarnings

* CLN address review comments

* Trigger [all random seeds] on the following tests:
test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

* CLN add comment for lbfgs ftol=64 * machine precision

* CLN XXX code comment

* Trigger [all random seeds] on the following tests:

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

* CLN link issue and remove code snippet in comment

* Trigger [all random seeds] on the following tests:

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

* CLN add catch_warnings

* Trigger [all random seeds] on the following tests:

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

* Trigger [all random seeds] on the following tests:

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

* [all random seeds]

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

* Trigger with -Werror [all random seeds]

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

* ENH increase maxls to 50

* [all random seeds]

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

* Revert "Trigger with -Werror [all random seeds]"

This reverts commit 99f4cf99ca41b4ad2bdad537ad60f936970e3a88.

* TST add catch_warnings to filterwarnings

* TST adapt tests for newton solvers

* CLN cleaner gradient step with gradient_times_newton

* DOC add whatsnew

* ENH always use lbfgs as fallback

* TST adapt rtol

* TST fix test_linalg_warning_with_newton_solver

* CLN address some review comments

* Improve tests related to convergence warning on collinear data

* overfit -> fit

* Typo in comment

* Apply suggestions from code review

* ENH fallback_lbfgs_solve
- Do not use lbfgs steps, fall back complete to lbfgs

* ENH adapt rtol

* Improve test_linalg_warning_with_newton_solver

* Better comments

* Fixed Hessian casing and improved warning messages

* [all random seeds]

test_linalg_warning_with_newton_solver

* Ignore ConvergenceWarnings for now if convergence is good

* CLN remove counting of warnings

* ENH fall back to lbfgs if line search did not converge

* DOC better comment on performance bottleneck

* Update GLM related examples to use the new solver

* CLN address reviewer comments

* EXA improve some wordings

* CLN do not pop "solver in parameter constraints

* CLN fix typos

* DOC fix docstring

* CLN remove solver newton-qr-cholesky

* DOC update PR number in whatsnew

* CLN address review comments

* CLN remove unnecessary catch_warnings

* CLN address some review comments

* DOC more precise whatsnew

* CLN use init_zero_coef

* CLN use and test init_zero_coef

* CLN address some review comments

* CLN mark NewtonSolver as private by leading underscore

* CLN exact comments for inner_solve

* TST add test_newton_solver_verbosity

* TST extend test_newton_solver_verbosity

* TST logic in test_glm_regression_unpenalized

* TST use count_nonzero

* CLN remove super rare line search checks

* MNT move Newton solver to new file _newton_solver.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>
Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

glemaitre committed:
MAINT force NumPy version for building scikit-learn for CPython 3.10 in Windows (#24742)

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>
Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

crflynn committed:
API add named_transformers attribute to FeatureUnion (#20331)


Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

lorentzenchr committed:
DOC fix deprecated log loss argument in user guide (#24753)

MaxwellLZH committed:
FIX check_estimator fails when validating SGDClassifier with log_loss (#24071)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

cmarmo committed:
DOC Add links to DBSCAN references. (#24758)

thomasjpfan committed:
FIX Fixes common test for requires_positive_X (#24667)

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

glemaitre committed:
DOC add entries for the 1.1.3 release (#24744)

glemaitre committed:
DOC update index showing new release

glemaitre committed:
DOC add more info about the drop of support for 32-bit Python on Windows  (#24761)

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

joaocmd committed:
DOC convert confusion matrix to y_true/y_pred for classification_report example (#24711)

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

thomasjpfan committed:
DOC Use show_config instead of numpy.distutils's get_info (#24762)

rprkh committed:
ENH Add dtype preservation for Isomap (#24714)

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

thomasjpfan committed:
FIX Improves nan support in LabelEncoder (#22629)

jeremiedbb commented:
I agree that it requires to be careful when interpreting these plots, but it's the case for everything. I don't think it's enough to just discard it.

Regarding the residuals plot, I proposed to add a histogram of the distribution on the side (https://github.com/scikit-learn/scikit-learn/pull/18020#issuecomment-1260657257), but I think it can be done in a later PR. It can also be another kind of aggregation as you mentioned.

Regarding "2. predicted vs actual", you are showing a pathological situation but I think it's still a valuable plot to have, at least for educational purpose because it's easy to explain and to understand what the plot represents.

Regarding the reliability diagram, I also agree that it's useful and that we should have it, but as said before it's still possible to extend the display later. We can start simple and add features afterwards. Discussing that irl with @ogrisel, he even thinks that it should have its own display `ReliabilityDiagramDisplay`. He also mentioned a `LorentzCurveDisplay` that would be interesting as well.

All that to say that I think it's important to have a simple display that is easy to explain and understand (yet not necessarily easy to interpret) because scikit-learn has an educational role. It also gives the opportunity to document the pros and cons of this kind of plots and guide more advanced users towards more interesting plots.

star1327p committed:
DOC changed marker colors for calibration comparison (#24766)

cmarmo committed:
DOC Fix inline interpreted text start-string without end-string. (#24734)

glevv committed:
ENH Add gamma='scale' option to RBFSampler (#24755)

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

fcharras committed:
TST Fix uniform rng in kmeans test_scaled_weights (#24778)

awinml committed:
DOC FIX Consistent formulae for metrics in the user guide (#24673)

Co-authored-by: Mike Altonji <mikealtonji@gmail.com>
Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

ArturoAmorQ reviewed (commented):
Thanks for the PR @glemaitre, here are some comments:

ggrrll committed:
Improve error message in _search_successive_halving.py (#24781)

MaxwellLZH committed:
FIX bagging with SGD and early stopping throws ZeroDivisionError (#23275)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

vnmabus committed:
ENH Allow path-like objects in load_svmlight_file. (#19075)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

adrinjalali committed:
DOC Update SECURITY.md for 1.1.3 (#24783)

glemaitre committed:
Update doc/modules/model_evaluation.rst

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

glemaitre committed:
Apply suggestions from code review

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

glemaitre committed:
Merge remote-tracking branch 'origin/main' into predicted_actual_regression_plot

glemaitre committed:
address missing comments from jeremie

glemaitre committed:
address remaining comments Arturo

glemaitre committed:
DOC fix suptitle

glemaitre committed:
rever tight_layout

lorentzenchr commented:
> I agree that it requires to be careful when interpreting these plots, but it's the case for everything. I don't think it's enough to just discard it.

As said earlier, I'd suggest to just start with the residual plot. This adds some value. The predicted-vs-actual is just a one-liner and its interpretation is usually flawed. A perfect model would be "around the diagonal", but not the diagonal itself. That, on the other hand, would only be true without randomness/noise.

GaelVaroquaux commented:
The predicted-vs-actual plot is very classic and used in many communities. People understand the meaning at a glance (even if the understanding is limited). I think that it is important that we support it.

mayer79 commented:
Coming from the statistics side, I'd back @lorentzenchr's position. 

A "residuals versus predicted" plot is a condensed version of all "residuals versus covariate" plots and thus provides a minimal check that a regression model is correctly specified. I'd call it the mother of all regression diagnostic plots.

On the other hand, a "predicted versus actual" plot often makes people think (wrongly) that their model is bad because the slope in the graph is not 1.  However, this is simply a result of regression to the mean, which itself is a consequence that the model cannot exactly predict Y|X, but rather E(Y|X).

jeremiedbb commented:
I don't know how to solve this debate :(
maybe some @scikit-learn/core-devs can help us decide ?

To summarize, this PR proposes a new display object for regression which can either plot the prediction error ("predicted vs actual") or the residuals ("predicted - actual vs predicted"). The debate is whether or not to include the first kind of plot. And it seems that the discussions may continue forever :)

Should this be solved by a vote ? How do we decide in the end if a feature should be included or not when there's no consensus ?

jjerphan commented:
Can't we just have both variation (i.e. a plot for predicted values vs. actual values and a plot for residuals)? To me, both one has benefits that the other doesn't and conversely.

GaelVaroquaux commented:
> Can't we just have both variation (i.e. a plot for predicted values vs. actual values and a plot for residuals)?

Yes, an argument to choose between one, the other or both. It seems the pragmatic view. Let's not be purists :)

adrinjalali commented:
I read @GaelVaroquaux 's comment as having a parameter defining which to include, which I would agree with.

ogrisel commented on doc/modules/model_evaluation.rst:
```diff
@@ -2711,6 +2711,45 @@ Here are some usage examples of the :func:`d2_absolute_error_score` function::
+predicts a specific quantile, e.g. using
+:class:`~sklearn.linear_model.QuantileRegressor`.
+
+The right-hand side plot shows the residuals, i.e. the difference between the
+actual values and the predicted values, vs. the predicted values. This type of
+plot give an indication of the `homoscedasticity
+<https://en.wikipedia.org/wiki/Homoscedasticity_and_heteroscedasticity>`_ of
+the model. A perfect regression model will display data points uniformly
+distributed around the horizontal axis. If the distribution is not uniform,
+then it indicates us that there is a relationship between the value of the
+target and the residuals, e.g. the error increases when the true target
+increases.
```
There is no reason for the "perfect model" to display homoschedastic residuals whenever the irreducible error is non-zero. If the irreducible error is null, then
the perfect model has zero residuals and the residuals homoschedastic as a consequence
but the homoschedasticity of a deterministic "random" variable is rather pointless.

Also the horizontal axis is not a density function so it does not make sense to speak about a uniform distribution when reading this plot... The distributions of residuals is rather to be "read vertically" for a given narrow, horizontal slice of `y_pred`.

Let' me suggest the following instead"
 
```suggestion
The right-hand side plot shows the residuals, i.e. the difference between the
actual values and the predicted values, vs. the predicted values.

This plot makes it easier to visualize if the residuals follow
and `homoscedastic or heteroschedastic
<https://en.wikipedia.org/wiki/Homoscedasticity_and_heteroscedasticity>`_
distribution. In particular, if the true distribution of `y|X` is Poisson
or Gamma distributed, it is expected that the variance of the residuals
of the optimal model to grow with the predicted value of `E[y|X]` (either linearly
for Poisson or quadratically for Gamma).

We can also use this plot to check if the residuals are Gaussian distributed
with a constant variance (homeschedastic residuals) which the assumption made
when fitting linear least squares regression model (see
:class:`sklearn.linear_mnodel.LinearRegression` and
:class:`sklearn.linear_mnodel.Ridge`). If this is not the case, and in particular if
the residual plot some banana-shaped structure, this is a hint that the model is
likely mis-specified and that non-linear feature engineering or switching to a
non-linear regression model might be useful.
```

ogrisel replied:
For reference, here is more information on how to interpret those kinds of plots:

https://www.qualtrics.com/support/stats-iq/analyses/regression-guides/interpreting-residual-plots-improve-regression/

I think it would be great to write some kind of advanced tutorial to diagnose various kinds of model misspecification problems for linear regression (penalized or not) but this should be done in a dedicated PR.

ogrisel commented on doc/whats_new/v1.2.rst:
```diff
@@ -442,6 +442,13 @@ Changelog
   of a binary classification problem. :pr:`22518` by
   :user:`Arturo Amor <ArturoAmorQ>`.
 
+- |Feature| Add :class:`metrics.PredictionErrorDisplay` to plot the predicted
+  and actual values to assess the performance of a regressor. The display can
+  be created with the class methods
```
```suggestion
- |Feature| Add :class:`metrics.PredictionErrorDisplay` to plot the predicted
  and actual values to qualitatively assess the behavior of a regressor. The
  display can be created with the class methods
```

ogrisel commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,409 @@
+        kind : {"predictions", "residuals"}, default="predictions"
+            The type of plot to draw:
+
+            - "predictions" draws the predicted values vs. the true values.
```
This should be true values (y axis) vs predict values (x axis). But we also need to check that this is actually what is done in the code.

ogrisel replied:
This does not seem to be the case when looking at the figures from the example included in the user guide.

glemaitre replied:
I searched widely and I could not find a clear consensus on plotting "predicted vs. actual" or "actual vs. predicted". Up to now, we always used "predicted (y-axis) vs. actual (x-axis)" in scikit-learn.

Since we actually have the same data, I would propose removing the parameter `x_axis` and going with "actual vs. predicted" to have always "predicted" in x-axis, as originally proposed by @lorentzenchr

lorentzenchr replied:
See Glen_b’s comment in https://stats.stackexchange.com/questions/104622/what-does-an-actual-vs-fitted-graph-tell-us
> It would be more common to follow the convention of plotting the values that are fixed (conditional on predictors) on the x-axis and the values that are random on the y-axis.

I had a hard time finding this plot in books at all😉

ogrisel commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,409 @@
+            - "predictions" draws the predicted values vs. the true values.
+            - "residuals" draws the residuals vs. the predicted values.
+
+        x_axis : {"actual_targets", "predicted_targets"}, default="actual_targets"
```
I would find it more natural to use `predicted_targets` as the default `x` axis, so that it's more natural to switch the "kind" argument without having to also change this argument.

ogrisel reviewed (commented):
I also agree that it's good to have actual vs predicted in addition of residuals vs predicted as it's a good educational opportunity to show that the residual plot leads to more informative interpretation (for instance to reveal model misspecification).

But we need to fix the explanation on how to read those plots. Let me suggest the following under the control of @lorentzenchr :)

+1 for including more options (e.g. heatmaps, various kinds of reliability diagrams) in follow up PRs. Either via new `kind` options for this class or via another display class if the parameters do not overlap.

glemaitre committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

glemaitre committed:
Merge remote-tracking branch 'origin/main' into predicted_actual_regression_plot

glemaitre committed:
iter

glemaitre committed:
iter

glemaitre committed:
iter

glemaitre committed:
iter

glemaitre commented:
I addressed the issue pointed out by @ogrisel

lorentzenchr commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,387 @@
+            Axes object to plot on. If `None`, a new figure and axes is
+            created.
+
+        kind : {"predictions", "residuals"}, default="predictions"
```
```suggestion
        kind : {"predictions", "residuals"}, default="residuals"
```

ogrisel replied:
Actually, I think we should rename `kind="predictions"` to `kind="actual_vs_predicted"`. It's more explicit even if a bit verbose. `kind="actual"` is ambiguous and `kind="predictions"` is a bit misleading because the name of a plot usually refers to what we put in the y-axis rather than the x-axis.

lorentzenchr commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,387 @@
+        self,
+        ax=None,
+        *,
+        kind="predictions",
```
```suggestion
        kind="residuals",
```

lorentzenchr commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,387 @@
+        X,
+        y,
+        *,
+        kind="predictions",
```
```suggestion
        kind="residuals",
```
Do we need this parameter at all? It is specified in `plot` again.

glemaitre replied:
Yes. Once the display is created, one can call a plot to get another type of plot because it uses the same data.

lorentzenchr commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,387 @@
+        y : array-like of shape (n_samples,)
+            Target values.
+
+        kind : {"predictions", "residuals"}, default="predictions"
```
```suggestion
        kind : {"predictions", "residuals"}, default="residuals"
```

lorentzenchr commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,387 @@
+        y_true,
+        y_pred,
+        *,
+        kind="predictions",
```
```suggestion
        kind="residuals",
```
See above.

lorentzenchr commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,387 @@
+        y_pred : array-like of shape (n_samples,)
+            Predicted target values.
+
+        kind : {"predictions","residuals"}, default="predictions"
```
```suggestion
        kind : {"predictions","residuals"}, default="residuals"
```

lorentzenchr reviewed (commented):
A partial review. Short on time atm.

ogrisel commented on doc/modules/model_evaluation.rst:
```diff
@@ -2711,6 +2711,70 @@ Here are some usage examples of the :func:`d2_absolute_error_score` function::
+arranged around the diagonal.
+
+Note that the above holds when the predicted values is the expected value of `y`
+given `X`. This is typically the case for regression model that asymptotically
```
> If we keep them, the log loss should be added.

This diagram is only valid when compute from outputs of scikit-learn regressors with continuous outcome variables. Residual plots for binary outcome variables seem uncommon/unexpected to me, and cannot be computed at all for multiclass outcomes.

Ok to drop the "asymptotically" for the sake of simplicity.

Also:

```suggestion
given `X`. This is typically the case for regression models that asymptotically
```

ogrisel commented on examples/model_selection/plot_cv_predict.py:
```diff
@@ -4,26 +4,65 @@
+# :class:`~sklearn.metrics.PredictionErrorDisplay` to visualize the
+# prediction errors.
+#
+# On the left axis, we plot the true values :math:`y` vs. the predicted values
```
I like "realized" and "observed". "actual" is already written on the y-axis of the plot so I think it's good to use a distinct synonym.

ogrisel commented:
I think that using `kind="residuals"` as the default is indeed a good idea because the class name `PredictionErrorDisplay` is about "prediction error" which better "residuals vs predicted" than "actual vs predicted".

ogrisel commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,387 @@
+                **line_kwargs,
+            )[0]
+            self.scatter_ = ax.scatter(
+                self.y_pred, self.y_pred - self.y_true, **scatter_kwargs
```
To make it easier to compare the point cloud of the "actual_vs_predicted" with the point cloud of "residuals", I think we should rather do:

```suggestion
                self.y_pred, self.y_true - self.y_pred, **scatter_kwargs
```

lorentzenchr replied:
Good catch. The standard residual is defined as suggested: r=y-prediction.

ogrisel commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,387 @@
+            self.scatter_ = ax.scatter(
+                self.y_pred, self.y_pred - self.y_true, **scatter_kwargs
+            )
+            xlabel, ylabel = "Predicted values", "Residuals"
```
```suggestion
            xlabel, ylabel = "Predicted values", "Residuals (actual - predicted)"
```

ogrisel commented on examples/model_selection/plot_cv_predict.py:
```diff
@@ -4,26 +4,65 @@
-# is a prediction obtained by cross validation:
-predicted = cross_val_predict(lr, X, y, cv=10)
+# %%
+# Since `cv=10`, it means that we trained 10 models and each of the model was
```
or even simpler: "each model".

ogrisel commented on examples/model_selection/plot_cv_predict.py:
```diff
@@ -4,26 +4,65 @@
 plt.show()
+
+# %%
+# It is important to note that
+# :func:`~sklearn.model_selection.cross_val_predict` is a tool that should be
+# used for visualization but not for evaluating the performance of the
+# predictive models by computing a metric on the aggregated predictions.
+#
+# A proper validation should use either
+# :func:`~sklearn.model_selection.cross_val_score` or
+# :func:`~sklearn.model_selection.cross_validate` instead.
```
```suggestion
# It is important to note that we used
# :func:`~sklearn.model_selection.cross_val_predict` for visualization
# purpose only in this example.
#
# It would be problematic to
# quantitatively assess the model performance by computing a single
# performance metric from the concatenated predictions returned by
# :func:`~sklearn.model_selection.cross_val_predict`
# when the different CV folds vary by size and distributions.
#
# In is recommended to compute per-fold performance metrics using:
# :func:`~sklearn.model_selection.cross_val_score` or
# :func:`~sklearn.model_selection.cross_validate` instead.
```

ogrisel commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,387 @@
+from ...utils import _safe_indexing
+
+
+class PredictionErrorDisplay:
+    """Prediction error visualization.
+
+    It is recommended to use
+    :func:`~sklearn.metrics.PredictionErrorDisplay.from_estimator` or
+    :func:`~sklearn.metrics.PredictionErrorDisplay.from_predictions` to
+    create a visualizer. All parameters are stored as attributes.
```
```suggestion
class PredictionErrorDisplay:
    """Visualization of the prediction error of a regression model.

    This tool can generate a "residual" or an "actual vs predicted" plot
    to qualitatively assess the behavior of a regressor, preferably on
    held out data points.

    See the details in the docstrings of
    :func:`~sklearn.metrics.PredictionErrorDisplay.from_estimator` or
    :func:`~sklearn.metrics.PredictionErrorDisplay.from_predictions` to
    create a visualizer. All parameters are stored as attributes.
```

ogrisel commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,387 @@
+    :func:`~sklearn.metrics.PredictionErrorDisplay.from_predictions` to
+    create a visualizer. All parameters are stored as attributes.
+
+    Read more in the :ref:`User Guide <visualizations>`.
```
we should also refer to the relevant section of the "model selection" chapter here.

ogrisel reviewed (commented):
More feedback.

Once this and @lorentzenchr's review comments are addressed, LGTM.

glemaitre committed:
Apply suggestions from code review

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

glemaitre committed:
DOC addressed most of the concerns

glemaitre committed:
FIX/TST respond reviews

glemaitre committed:
fix

ogrisel commented on doc/modules/model_evaluation.rst:
```diff
@@ -2711,6 +2711,80 @@ Here are some usage examples of the :func:`d2_absolute_error_score` function::
+the data at hand. This plot is therefore useful for making a choice between
+different family of models.
+
+We can also use this plot to check if the residuals are Gaussian distributed
+with a constant variance (homeschedastic residuals) which is the assumption
+made when fitting linear least squares regression model (see
+:class:`sklearn.linear_mnodel.LinearRegression` and
+:class:`sklearn.linear_mnodel.Ridge`). If this is not the case, and in
+particular if the residual plot some banana-shaped structure, this is a hint
+that the model is likely mis-specified and that non-linear feature engineering
+or switching to a non-linear regression model might be useful.
```
```suggestion
When fitting a linear least squares regression model (see :class:`sklearn.linear_mnodel.LinearRegression` and
:class:`sklearn.linear_mnodel.Ridge`), we can use this plot to check
if some of the `model assumptions
<https://en.wikipedia.org/wiki/Ordinary_least_squares#Assumptions>`_
are met, in particular that the residuals should be uncorrelated, their
expected value should be null and that their variance should be constant
(homoschedasticity).

If this is not the case, and in particular if the residual plot some
banana-shaped structure, this is a hint that the model is likely mis-specified
and that non-linear feature engineering or switching to a non-linear regression
model might be useful.
```

ogrisel replied:
Do you agree with the above suggestion @lorentzenchr?

ogrisel replied:
I do not necessarily want to go into more details about estimator efficiency. I think people can go read the wikipedia article that I linked which also links to the article about the Gauss Markov Theorem on one hand and on generalized least squares on the other hand (for correlated residuals with a known correlation structure).

ogrisel commented on doc/modules/model_evaluation.rst:
```diff
@@ -2711,6 +2711,80 @@ Here are some usage examples of the :func:`d2_absolute_error_score` function::
+
+In particular, if the true distribution of `y|X` is Poisson or Gamma
+distributed, it is expected that the variance of the residuals of the optimal
+model to grow with the predicted value of `E[y|X]` (either linearly for
```
```suggestion
model would grow with the predicted value of `E[y|X]` (either linearly for
```

ogrisel commented on doc/modules/model_evaluation.rst:
```diff
@@ -2711,6 +2711,80 @@ Here are some usage examples of the :func:`d2_absolute_error_score` function::
+model to grow with the predicted value of `E[y|X]` (either linearly for
+Poisson or quadratically for Gamma).
+
+Since a predictive models do not adapt themselves to the particularities of the
+true distribution, it comes to us at choosing the optimal modelisation given
+the data at hand. This plot is therefore useful for making a choice between
+different family of models.
+
```
I don't think the first sentence is true. Any non-trivial machine learning model
adapts its parameters to best match the empirical distribution.

But even if you refer to shape of the true conditional distribution and try to match it by choosing different fitting criterions or parametrizations, it would still be possible to design a metaestimator that runs and an internal CV and tries many estimator classes.

Anyways, I don't think this paragraph brings any useful information beyond what is explained in more specific previous and following paragraphs. Instead it can bring confusion so I would suggest to just drop it.

```suggestion
```

ogrisel commented on doc/modules/model_evaluation.rst:
```diff
@@ -2711,6 +2711,80 @@ Here are some usage examples of the :func:`d2_absolute_error_score` function::
+
+Note that the above only holds when the predicted values is the expected value
+of `y` given `X`. This is typically the case for regression models that
+asymptotically minimize the mean squared error objective function or more
+generally the :ref:`mean Tweedie deviance <mean_tweedie_deviance>` for any
+value of its "power" parameter.
```
Lets drop the "asymptotically" as discussed earlier:

```suggestion
minimize the mean squared error objective function or more generally the
:ref:`mean Tweedie deviance <mean_tweedie_deviance>` for any value of its
"power" parameter.
```

ogrisel commented on doc/modules/model_evaluation.rst:
```diff
@@ -2711,6 +2711,80 @@ Here are some usage examples of the :func:`d2_absolute_error_score` function::
+
+When plotting the predictions of an estimator that predicts a quantile
+of `y` given `X`, e.g. :class:`~sklearn.linear_model.QuantileRegressor`
+or any other model asymptotically minimizing the :ref:`pinball loss
+<pinball_loss>`, a fraction of the points are either expected to lie above or
+below the diagonal depending on the estimated quantile level.
```
Same:

```suggestion
or any other model minimizing the :ref:`pinball loss <pinball_loss>`, a
fraction of the points are either expected to lie above or below the diagonal
depending on the estimated quantile level.
```

ogrisel commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,403 @@
+
+            - "actual_vs_predicted" draws the the observed values (y-axis) vs.
+              the predicted values (x-axis).
+            - "residuals" draws the residuals, i.e difference between observed
+              and predicted values, (y-axis) vs. the predicted values (x-axis).
```
```suggestion
            - "residual_vs_predicted" draws the residuals, i.e difference
              between observed and predicted values, (y-axis) vs. the predicted
              values (x-axis).
```

ogrisel reviewed (approved):
Final pass. Other than that and the previous suggestion, LGTM.

glemaitre committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

glemaitre committed:
DOC apply same changes to other docstrings

glemaitre committed:
Merge remote-tracking branch 'origin/main' into predicted_actual_regression_plot

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/glemaitre/18020

jeremiedbb committed:
fix what's new

ogrisel commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,406 @@
+    For general information regarding `scikit-learn` visualization tools, read
+    more in the :ref:`Visualization Guide <visualizations>`.
+    For details regarding interpreting these plots, refer to the
+    :ref`Model Evaluation Guide <visualization_regression_evaluation>`.
```
![image](https://user-images.githubusercontent.com/89061/203957673-1c1b939e-5d2a-49e5-80a7-464b9be7615d.png)

ogrisel replied:
```suggestion
    :ref:`Model Evaluation Guide <visualization_regression_evaluation>`.
```

ogrisel commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,406 @@
+        For general information regarding `scikit-learn` visualization tools,
+        read more in the :ref:`Visualization Guide <visualizations>`.
+        For details regarding interpreting these plots, refer to the
+        :ref`Model Evaluation Guide <visualization_regression_evaluation>`.
```
```suggestion
        :ref:`Model Evaluation Guide <visualization_regression_evaluation>`.
```

ogrisel commented on sklearn/metrics/_plot/regression.py:
```diff
@@ -0,0 +1,406 @@
+        For general information regarding `scikit-learn` visualization tools,
+        read more in the :ref:`Visualization Guide <visualizations>`.
+        For details regarding interpreting these plots, refer to the
+        :ref`Model Evaluation Guide <visualization_regression_evaluation>`.
```
```suggestion
        :ref:`Model Evaluation Guide <visualization_regression_evaluation>`.
```

ogrisel committed:
Apply suggestions from code review

jeremiedbb committed:
fix broken links

jeremiedbb commented:
Let's go !

We can still improve some wordings later if needed.

Thanks @glemaitre !
